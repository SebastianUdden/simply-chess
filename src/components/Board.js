import React, { useState, useEffect } from "react"
import styled from "styled-components"
import { CHESS_BOARD, CHESS_COLUMNS } from "../constants/chessBoard"
import { MOVEMENT_PATTERN } from "../constants/movementPattern"
import { checkIsValid, showPossibleMoves } from "../utils/validateMovement"
import Cell from "./Cell"

const Board = styled.div`
  margin: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column-reverse;
  justify-content: space-between;
`
const Row = styled.div`
  display: flex;
  justify-content: space-between;
  height: 100%;
`
const Title = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  color: orange;
  border: none;
  text-transform: capitalize;
  ${p =>
    p.minWidth &&
    `
    min-width: 40px;
    min-height: 25px;
    `}
`

const isEnemy = (clickedPiece, selectedPiece) =>
  clickedPiece &&
  selectedPiece &&
  selectedPiece.id !== clickedPiece.id &&
  selectedPiece.color !== clickedPiece.color

const cellMatches = (selection, cell, row) =>
  row.number === selection.row && cell.name === selection.cell.name

const DEFAULT_SELECTED = {
  row: "",
  cell: { name: "", symbol: undefined },
}

export default ({ playerTurn, onPlayerMove, onClose }) => {
  if (!CHESS_BOARD) return null
  if (!CHESS_BOARD.rows) return null

  const [selectedCell, setSelectedCell] = useState(DEFAULT_SELECTED)
  const [board, setBoard] = useState(CHESS_BOARD)

  const handleSelectCell = newSelection => {
    const selectedPiece = selectedCell.cell.symbol
    const clickedPiece = newSelection.cell.symbol
    const enemyCell = isEnemy(clickedPiece, selectedPiece)
    const emptyCell = !newSelection.cell.symbol
    const validPosition =
      (enemyCell || emptyCell) &&
      checkIsValid(newSelection, selectedCell, board, MOVEMENT_PATTERN)

    const getCellValue = (cell, row) => {
      if (validPosition && enemyCell && clickedPiece.type === "king") {
        setTimeout(() => onClose(selectedPiece.color), 500)
      }
      const symbol = validPosition ? selectedPiece : cell.symbol
      if (cellMatches(newSelection, cell, row)) {
        const selected = validPosition ? false : true
        return {
          ...cell,
          selected: selected && cell.symbol && cell.symbol.color === playerTurn,
          symbol,
        }
      }
      if (cellMatches(selectedCell, cell, row)) {
        return {
          ...cell,
          selected: false,
          symbol: validPosition ? undefined : cell.symbol,
        }
      }
      return { ...cell, selected: false }
    }

    setBoard({
      rows: board.rows.map(row => ({
        ...row,
        cells: row.cells.map(cell => getCellValue(cell, row)),
      })),
    })
    if (validPosition) {
      onPlayerMove(selectedCell.cell.symbol.color)
      setSelectedCell(DEFAULT_SELECTED)
      return
    }
    if (
      newSelection.cell &&
      newSelection.cell.symbol &&
      newSelection.cell.symbol.color === playerTurn
    ) {
      setSelectedCell(newSelection)
    } else {
      setSelectedCell(DEFAULT_SELECTED)
    }
  }

  useEffect(() => {
    if (!board) return
    if (!selectedCell) return
    setBoard(showPossibleMoves(selectedCell, board, MOVEMENT_PATTERN))
  }, [selectedCell])

  return (
    <Board>
      <Row>
        <Title></Title>
        {CHESS_COLUMNS.map(column => (
          <Title key={column.name} minWidth>
            {column.name}
          </Title>
        ))}
        <Title></Title>
      </Row>
      {board.rows.map(row => (
        <Row key={row.number}>
          <Title>{row.number}</Title>
          {row.cells.map(cell => (
            <Cell
              key={row.number + cell.name}
              cell={cell}
              row={row}
              onClick={() => handleSelectCell({ row: row.number, cell })}
            />
          ))}
          <Title></Title>
        </Row>
      ))}
      <Row>
        <Title></Title>
        {board.rows.map(row => (
          <Title key={row.number}></Title>
        ))}
        <Title></Title>
      </Row>
    </Board>
  )
}
