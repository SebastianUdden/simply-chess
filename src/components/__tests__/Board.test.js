import React from "react"
import renderer from "react-test-renderer"
import Board from "../Board"

describe("Board", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Board playerTurn="white" onPlayerMove={() => {}} onClose={() => {}} />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
