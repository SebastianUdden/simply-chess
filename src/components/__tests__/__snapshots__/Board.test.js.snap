// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders correctly 1`] = `
<div
  className="sc-AxiKw kFENUr"
>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy ixzmtH"
    >
      A
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      B
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      C
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      D
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      E
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      F
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      G
    </div>
    <div
      className="sc-AxhUy ixzmtH"
    >
      H
    </div>
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      1
    </div>
    <div
      className="sc-AxjAm fNZKtw"
      color="#113111"
      data-cy=""
      id="A1"
      onClick={[Function]}
    >
      <img
        alt="tower"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm ekbssn"
      color="#114111"
      data-cy=""
      id="B1"
      onClick={[Function]}
    >
      <img
        alt="knight"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="C1"
      onClick={[Function]}
    >
      <img
        alt="runner"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="D1"
      onClick={[Function]}
    >
      <img
        alt="queen"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="E1"
      onClick={[Function]}
    >
      <img
        alt="king"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="F1"
      onClick={[Function]}
    >
      <img
        alt="runner"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="G1"
      onClick={[Function]}
    >
      <img
        alt="knight"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="H1"
      onClick={[Function]}
    >
      <img
        alt="tower"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      2
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="A2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="B2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="C2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="D2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="E2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="F2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="G2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="H2"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      3
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="A3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="B3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="C3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="D3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="E3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="F3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="G3"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="H3"
      onClick={[Function]}
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      4
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="A4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="B4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="C4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="D4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="E4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="F4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="G4"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="H4"
      onClick={[Function]}
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      5
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="A5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="B5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="C5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="D5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="E5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="F5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="G5"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="H5"
      onClick={[Function]}
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      6
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="A6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="B6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="C6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="D6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="E6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="F6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="G6"
      onClick={[Function]}
    />
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="H6"
      onClick={[Function]}
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      7
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="A7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="B7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="C7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="D7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="E7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="F7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="G7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="H7"
      onClick={[Function]}
    >
      <img
        alt="pawn"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    >
      8
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="A8"
      onClick={[Function]}
    >
      <img
        alt="tower"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="B8"
      onClick={[Function]}
    >
      <img
        alt="knight"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="C8"
      onClick={[Function]}
    >
      <img
        alt="runner"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="D8"
      onClick={[Function]}
    >
      <img
        alt="queen"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="E8"
      onClick={[Function]}
    >
      <img
        alt="king"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="F8"
      onClick={[Function]}
    >
      <img
        alt="runner"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm OnsRP"
      color="#555"
      data-cy=""
      id="G8"
      onClick={[Function]}
    >
      <img
        alt="knight"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxjAm bbaRbW"
      color="#262626"
      data-cy=""
      id="H8"
      onClick={[Function]}
    >
      <img
        alt="tower"
        className="sc-AxirZ huvrLf"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
  <div
    className="sc-AxhCb fyhQQo"
  >
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
    <div
      className="sc-AxhUy gKCKPz"
    />
  </div>
</div>
`;
